### PR TITLE
Refactor memory and disk storages

### PR DESCRIPTION
**Pull request by vadymmarkov**
This PR refactors `DiskCache` and `MemoryCache` to be sync, all the async logic is handled in `BasicHybridCache` instead. It improves code readability and makes it easier to test.

### CODE REVIEW

File: Source/Shared/Cache/SpecializedCache.swift [hunk 1]
```diff
@@ -13,25 +13,25 @@ public final class SpecializedCache<T: Cachable>: BasicHybridCache {
    - Parameter expiry: Expiration date for the cached object
    - Parameter completion: Completion closure to be called when the task is done
    */
-  public func add(_ key: String, object: T, expiry: Expiry? = nil, completion: (() -> Void)? = nil) {
-    super.add(object, forKey: key, expiry: expiry, completion: completion)
+  public func add(_ key: String, object: T, expiry: Expiry? = nil, completion: Completion? = nil) {
+    return super.add(object, forKey: key, expiry: expiry, completion: completion)
```
Comment (onmyway133) [inline Source/Shared/Cache/SpecializedCache.swift]: It just calls `super` ?

File: Source/Shared/Config.swift [hunk 2]
```diff
@@ -7,8 +7,10 @@ public struct Config {
   /// Expiry date that will be applied by default for every added object
   /// if it's not overridden in the add(key: object: expiry: completion:) method
   public let expiry: Expiry
-  /// Maximum amount of items to store in memory
-  public let maxObjectsInMemory: Int
+  /// The maximum number of objects the cache should hold
```
Comment (onmyway133) [inline Source/Shared/Config.swift]: `of objects in memory`?

**Comment by onmyway133** [comment]
I think this should simplify things 🚀

File: Source/Shared/Cache/SpecializedCache.swift (hunk 1 above)
Comment (vadymmarkov) [inline Source/Shared/Cache/SpecializedCache.swift]: Yes, but `SpecializedCache` is generic and superclass is not, so it does the trick.

File: Source/Shared/Cache/BasicHybridCache.swift [hunk 3]
```diff
@@ -77,186 +96,214 @@ public class BasicHybridCache: NSObject {
...
-
-      weakSelf.backStorage.clearExpired {
-        completion?()
+  func cacheEntry<T: Cachable>(forKey key: String, completion: @escaping (CacheEntry<T>?) -> Void) {
+    readQueue.async { [weak self] in
+      do {
+        guard let `self` = self else {
+          throw CacheError.notFound
+        }
+        if let entry: CacheEntry<T> = self.frontStorage.cacheEntry(key) {
```
Comment (onmyway133) [inline Source/Shared/Cache/BasicHybridCache.swift]: So much cleaner

File: Source/Shared/Cache/BasicHybridCache.swift [hunk 4]
```diff
@@ -9,6 +9,12 @@
  Cachable protocol. It's two layered cache (with front and back storages)
  */
 public class BasicHybridCache: NSObject {
+  enum CacheError: Error {
```
Comment (onmyway133) [inline Source/Shared/Cache/BasicHybridCache.swift]: Or we can just name it `Error`?

Comment (vadymmarkov) [inline Source/Shared/Cache/BasicHybridCache.swift]: Then I have to use `Swift.Error` in completion closures, and I don't really like it.

**Pull request merged**